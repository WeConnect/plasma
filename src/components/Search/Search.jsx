// @flow
import React, {
  Component,
  type Node,
} from 'react';
import cx from 'classnames';
import Autocomplete from 'react-autocomplete';

import { getDataAttrs } from '../../dataUtils';
import type { Data } from '../../types';
import styles from './style.scss';

type Hash<V> = { [key: string]: V };

type Item = {| value: string, label: string |} & Hash<any>;

type Props = {|
  autoHighlight?: boolean,
  clearable: boolean,
  clearIcon: Node,
  data: Data,
  disabled: boolean,
  getItemValue: (item: Item) => string,
  icon: Node,
  isItemSelectable: (item: Item) => boolean,
  items: Array<Item>,
  onChange: (value: string) => void,
  onClear: () => void,
  onMenuVisibilityChange: (isOpen: boolean) => void,
  onSelect?: (value: string, item: Item) => void,
  open?: boolean,
  placeholder: string,
  selectOnBlur?: boolean,
  shouldItemRender?: (item: Item, value: string) => boolean,
  sortItems?: (itemA: Item, itemB: Item, value: string) => number,
  renderInput?: (props: Hash<any>) => Node,
  renderItem: (item: Item, isHighlighted: boolean, styles: Hash<any>) => Node,
  renderMenu?: (items: Array<Node>, value: string, styles: Hash<any>) => Node,
  value?: string,
|};

type State = {|
  value: ?string,
|};

class Search extends Component<Props, State> {
  static defaultProps = {
    clearable: true,
    clearIcon: <span className={styles.defaultClearIcon}>&times;</span>,
    getItemValue: (item: Item): string => item.value,
    renderItem: (item: Item, isHighlighted: boolean): Node => {
      const classes = cx(styles.defaultItem, {
        [styles.defaultItemHighlighted]: isHighlighted,
      });

      return <div key={item.value} className={classes}>{item.label}</div>;
    },
    renderMenu: (children: Array<Node>, value: string): Node => {
      const showEmpty = value && !children.length;

      return (
        <div className={styles.defaultMenu}>
          {
            showEmpty ?
              (
                <div className={cx(styles.defaultItem, styles.defaultItemEmpty)}>
                  No results found
                </div>
              ) :
              children
          }
        </div>
      );
    },
    shouldItemRender(item: Item, value: string): boolean {
      return item.label.toLowerCase().indexOf(value.toLowerCase()) !== -1;
    },
  };

  state = {
    value: this.props.value,
  };

  componentWillReceiveProps(nextProps: Props) {
    const { value } = nextProps;

    if (value !== this.props.value) {
      this.setState({ value });
    }
  }

  handleChange = (evt: SyntheticKeyboardEvent<HTMLInputElement>, value: string): void => {
    const {
      onChange,
    } = this.props;

    this.setState({ value: evt.target.value });

    onChange && onChange(value);
  };

  handleSelect = (value: string, item: Item): void => {
    const {
      onSelect,
    } = this.props;

    this.setState({
      value,
    });

    onSelect && onSelect(value, item);
  };

  handleClear = (): void => {
    const {
      onClear,
    } = this.props;

    this.setState({ value: '' });

    onClear && onClear();
  };

  render() {
    const {
      clearable,
      clearIcon,
      data,
      disabled,
      icon,
      placeholder,
      ...rest
    } = this.props;

    const showClearIcon = clearable && this.state.value;
    const showInputIcon = !!icon;

    const wrapperProps = {
      ...getDataAttrs(data),
      style: {},
      className: styles.inputWrapper,
    };

    const inputProps = {
      className: cx(styles.input, {
        [styles.inputHasIcon]: showInputIcon,
        [styles.inputHasClearIcon]: showClearIcon,
      }),
      disabled,
      placeholder,
    };

    return (
      <div
        {...getDataAttrs(data)}
        className={styles.container}
      >
        {
          showInputIcon && (
            <div className={cx(styles.iconContainer, styles.inputIconContainer)}>
              {icon}
            </div>
          )
        }
        {
          showClearIcon && (
            <button
              className={cx(styles.iconContainer, styles.clearableIconContainer)}
              onClick={this.handleClear}
            >
              {clearIcon}
            </button>
          )
        }
        <Autocomplete
          {...rest}
          wrapperProps={wrapperProps}
          inputProps={inputProps}
          value={this.state.value}
          onChange={this.handleChange}
          onSelect={this.handleSelect}
        />
      </div>
    );
  }
}

Search.displayName = 'Plasma@Search';

export default Search;
