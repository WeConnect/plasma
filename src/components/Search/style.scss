@import '../../styles/base';

.container {
  @extend %type--body;

  position: relative;
  display: flex;
  align-items: center;
}
.inputWrapper {
  width: 100%;
  height: 40px;
}

.input {
  @extend %input;
  @extend %input--wrapper;

  outline: 0;
  padding: 5px 10px;

  &:hover {
    @extend %input--wrapper--hover;
  }

  &:focus {
    @extend %input--wrapper--focus;
  }

  &::placeholder {
    color: $cGray200;
  }

  &:disabled,
  &:disabled:hover {
    @extend %input--wrapper--disabled;

    &::placeholder {
      color: $cGray300;
    }
  }

  &:invalid,
  &:invalid:hover {
    @extend %input--wrapper--error;
  }

  &.inputHasIcon {
    padding-left: 50px;
  }

  &.inputHasClearIcon {
    padding-right: 40px;
  }
}

.searchIconContainer {
  position: absolute;
  display: flex;
  width: 50px;
  height: 100%;
  align-items: center;
  justify-content: center;
}

.iconContainer {
  position: absolute;
}
.inputIconContainer {
  align-items: center;
  display: flex;
  justify-content: center;
  left: 10px;
  width: 30px;
  height: 30px;
}
.clearableIconContainer {
  border: 0;
  background: transparent;
  cursor: pointer;
  line-height: 20px;
  outline: 0;
  padding: 0;
  right: 10px;
  vertical-align: middle;
  width: 20px;
  height: 20px;

  &:hover + .inputWrapper .input,
  &:active + .inputWrapper .input {
    @extend %input--wrapper--hover;
  }
}

.clearableIconContainer {
  .defaultClearIcon {
    color: $cGray300;
    font-size: 18px;
    font-weight: bold;
  }

  &:hover .defaultClearIcon {
    color: $cGray200;
  }

  &:active .defaultClearIcon {
    color: $cGray60;
  }
}

.defaultMenu {
  @extend %dropdown;

  cursor: pointer;
}

.defaultItem {
  @extend %dropdown-item;

  &.defaultItemHighlighted {
    @extend %dropdown-item--focus;
  }
  &.defaultItemEmpty {
    color: $cGray200;
    padding-top: 10px;
    padding-bottom: 10px;
  }
}
