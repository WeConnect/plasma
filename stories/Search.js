import React from 'react';
import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import { linkTo } from '@storybook/addon-links';
import Search from '../src/components/Search/Search';

storiesOf('Search', module).add('default', () => {
  return <Search iconUrl="http://placehold.it/30x30" onChange={() => {}} placeholder="Search" />;
});

storiesOf('Search', module).add('with a small icon passed in', () => {
  return <Search iconUrl="http://placehold.it/16x16" onChange={() => {}} placeholder="Search" />;
});

storiesOf('Search', module).add('with instruction text', () => {
  return (
    <Search
      iconUrl="http://placehold.it/30x30"
      instructionText="Some sort of instructions for this search."
      onChange={() => {}}
      placeholder="Search"
    />
  );
});

storiesOf('Search', module).add('clearable', () => {
  return (
    <Search
      clearable
      clearIconUrl="http://placehold.it/8x8"
      iconUrl="http://placehold.it/30x30"
      onChange={() => {}}
      placeholder="Search"
    />
  );
});

storiesOf('Search', module).add('disabled', () => {
  return (
    <Search disabled iconUrl="http://placehold.it/30x30" onChange={() => {}} placeholder="Search" />
  );
});

storiesOf('Search', module).add('with prepopulated value', () => {
  return (
    <Search
      clearable
      clearIconUrl="http://placehold.it/8x8"
      iconUrl="http://placehold.it/30x30"
      onChange={() => {}}
      placeholder="Search"
      value="Hello there"
    />
  );
});

storiesOf('Search', module).add('with empty prepopulated value', () => {
  return (
    <Search
      clearable
      clearIconUrl="http://placehold.it/8x8"
      iconUrl="http://placehold.it/30x30"
      onChange={() => {}}
      placeholder="Search"
      value=""
    />
  );
});
