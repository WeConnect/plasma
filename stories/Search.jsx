import React from 'react';
import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';

import Search from '../src/components/Search/Search';

const items = 'ABCDEF'
  .split('')
  .map(c => ({
    label: c + c + c,
    value: c + c + c,
  }));

const defaultProps = {
  items,
  onChange: action('changed'),
  placeholder: 'Search',
};

const stories = storiesOf('Search', module);

stories.add('default', () => <Search {...defaultProps} />);

stories.add('with an icon', () => (
  <Search
    {...defaultProps}
    icon={<img src="http://placehold.it/30x30" alt="" />}
  />
));
stories.add('with a small icon', () => (
  <Search
    {...defaultProps}
    icon={<img src="http://placehold.it/16x16" alt="" />}
  />
));
stories.add('clearable', () => (
  <Search
    {...defaultProps}
    clearable
    value="Some value"
  />
));
stories.add('disabled', () => (
  <Search
    {...defaultProps}
    disabled
  />
));
stories.add('custom clear icon', () => (
  <Search
    {...defaultProps}
    clearable
    clearIcon={<img src="http://placehold.it/20x20" alt="" />}
    value="Some value"
  />
));
