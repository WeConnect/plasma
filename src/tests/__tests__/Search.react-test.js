/* eslint-disable */

import React from 'react';
import Search from '../../components/Search/Search';
import renderer from 'react-test-renderer';
import { shallow, mount } from 'enzyme';

describe('Search Component', () => {
  test('render', () => {
    const component = renderer.create(
      <Search
        items={[]}
        icon={<img src="http://yolo.com/icon.png" />}
      />
    );
    expect(component).toMatchSnapshot();
  });

  test('render value', () => {
    const component = renderer.create(
      <Search
        items={[]}
        clearable
        clearIcon={<img src="http://yolo.com/icon.png" />}
        value="test"
      />
    );

    expect(component).toMatchSnapshot();
  });

  test('onClear', () => {
    const onClear = jest.fn();
    const component = mount(
      <Search
        items={[]}
        clearable
        clearIcon={<img src="http://yolo.com/icon.png" />}
        onClear={onClear}
      />
    );
    component.setState({ value: 'text' });
    component.find('button').simulate('click');
    expect(onClear).toHaveBeenCalledTimes(1);
  });

  test('onChange', () => {
    const onChange = jest.fn();
    const component = mount(
      <Search
        items={[]}
        onChange={onChange}
      />
    );

    const input = component.find('input');

    input.getDOMNode().value = 'YOLO';
    input.simulate('change');

    expect(onChange).toHaveBeenCalledTimes(1);
    expect(component.state('value')).toBe('YOLO');
  });
});
