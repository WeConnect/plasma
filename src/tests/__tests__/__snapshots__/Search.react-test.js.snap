// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search Component render 1`] = `
<div
  className="container"
>
  <div
    className="iconContainer inputIconContainer"
  >
    <img
      src="http://yolo.com/icon.png"
    />
  </div>
  <div
    className="inputWrapper"
    style={Object {}}
  >
    <input
      aria-autocomplete="list"
      aria-expanded={false}
      autoComplete="off"
      className="input inputHasIcon"
      disabled={undefined}
      onBlur={[Function]}
      onChange={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      placeholder={undefined}
      role="combobox"
      value=""
    />
  </div>
</div>
`;

exports[`Search Component render value 1`] = `
<div
  className="container"
>
  <button
    className="iconContainer clearableIconContainer"
    onClick={[Function]}
  >
    <img
      src="http://yolo.com/icon.png"
    />
  </button>
  <div
    className="inputWrapper"
    style={Object {}}
  >
    <input
      aria-autocomplete="list"
      aria-expanded={false}
      autoComplete="off"
      className="input inputHasClearIcon"
      disabled={undefined}
      onBlur={[Function]}
      onChange={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      placeholder={undefined}
      role="combobox"
      value="test"
    />
  </div>
</div>
`;
